python/pyarrow/_compute.pyx

```diff
@@ -2629,14 +2718,98 @@ def register_scalar_function(func, function_name, function_doc, in_types,
       21
     ]
     """
+    return register_scalar_like_function(GetRegisterScalarFunction(),
+                                         func, function_name, function_doc, in_types,
+                                         out_type, func_registry)
+
+
+def register_tabular_function(func, function_name, function_doc, in_types, out_type,
+                              func_registry=None):
+    """
+    Register a user-defined tabular function.
+
+    A tabular function is one accepting a context argument of type
+    ScalarUdfContext and returning a generator of struct arrays.
+    The in_types argument must be empty and the out_type argument
+    specifies a schema. Each struct array must have field types
+    correspoding to the schema.
+
+    Parameters
+    ----------
+    func : callable
+        A callable implementing the user-defined function.
+        The only argument is the context argument of type
+        ScalarUdfContext. It must return a callable that
+        returns on each invocation a StructArray matching
+        the out_type, where an empty array indicates end.
+    function_name : str
+        Name of the function. This name must be globally unique.
+    function_doc : dict
+        A dictionary object with keys "summary" (str),
+        and "description" (str).
+    in_types : Dict[str, DataType]
+        Must be an empty dictionary.
+    out_type : DataType
+        Output type of the function.
+    func_registry : FunctionRegistry
+        Optional function registry to use instead of the default global one.
+    """
+    return register_scalar_like_function(GetRegisterTabularFunction(),
+                                         func, function_name, function_doc, in_types,
+                                         out_type, func_registry)
+
+
+def register_scalar_like_function(register_func, func, function_name, function_doc, in_types,
```

Comment:
the table functions don't take any `in_types`, so I don't see why it needs to share that logic with scalar like. Perhaps we should simply add a utility function to handle the common doc validation for UDF registration?

Reply:
The design is that they will take inputs, but only no-inputs is currently implemented. See [this post](https://github.com/apache/arrow/pull/14682#issuecomment-1322610029).

Reply:
I'll fix to `NotImplemented` in this case.